examples/xlio_ultra_api_ping_pong.c

```diff
@@ -822,115 +697,100 @@ static void socket_accept_cb(xlio_socket_t sock, xlio_socket_t parent, uintptr_t
+        /* Keep doubling until it fits */
+        while (new_capacity < acc->len + len) {
+            new_capacity *= 2;
+        }
 
-    if (!g_state.mr_buf) {
-        fprintf(stderr, "Memory not registered for zero-copy\n");
-        return -1;
-    }
+        /* Allocate or reallocate buffer (realloc handles NULL first argument) */
+        new_data = realloc(acc->data, new_capacity);
+        if (!new_data) {
+            fprintf(stderr, "Failed to %s RX accumulator buffer to %zu bytes\n",
+                    acc->data ? "grow" : "allocate", new_capacity);
+            return;
+        }
 
-    if (len >= sizeof(g_state.sndbuf)) {
-        fprintf(stderr, "Message too long for send buffer\n");
-        return -1;
+        acc->data = new_data;
+        acc->capacity = new_capacity;
     }
 
-    /* Copy message to registered memory */
-    memcpy(g_state.sndbuf, msg, len);
+    /* Step 2: Append new data to accumulator */
+    memcpy(acc->data + acc->len, data, len);
+    acc->len += len;
 
-    /*
-     * Configure send attributes for zero-copy operation
-     * ================================================
-     */
-    struct xlio_socket_send_attr send_attr = {
-        .flags = XLIO_SOCKET_SEND_FLAG_FLUSH, /* Flush immediately */
-        .mkey = g_state.mr_buf->lkey, /* Memory key for zero-copy */
-        .userdata_op = (uintptr_t)msg /* User data for completion callback */
-    };
+    /* Step 3: Process all complete messages in the accumulator */
+    msg_start = acc->data;
+    while (acc->len > 0) {
+        /* Search for the message delimiter in current accumulated data */
+        newline_pos = memchr(msg_start, '\n', acc->len);
+        if (!newline_pos) {
+            break;
+        }
 
-    /*
-     * Send data using XLIO Ultra API
-     * ==============================
-     */
-    int rc = g_state.api->xlio_socket_send(sock, g_state.sndbuf, len, &send_attr);
-    if (rc != 0) {
-        fprintf(stderr, "Failed to send message: %s\n", strerror(errno));
-        return rc;
-    }
+        /* Found a complete message */
+        msg_len = newline_pos - msg_start;
+        /* Make a null-terminate message in-place */
+        *newline_pos = '\0';
 
-    return 0;
-}
+        printf("Received message: '%s' (%zu bytes)\n", msg_start, msg_len);
 
-/*
- * Send ping message with RTT timing
- * =================================
- * Records timestamp when sending ping for RTT measurement.
- */
-static int send_ping_with_timing(xlio_socket_t sock)
-{
-    /* Record timestamp before sending ping */
-    if (clock_gettime(CLOCK_MONOTONIC, &g_state.ping_sent_time) != 0) {
-        fprintf(stderr, "Failed to get timestamp: %s\n", strerror(errno));
-        return -1;
+        /* Process the null-terminated message */
+        process_received_message(sock, msg_start, msg_len);
+
+        /* Move to the next message */
+        msg_start = newline_pos + 1;
```

Comment:
suppose we get:
"PING\nPONG\n" (len = 10)

iteration 1 - acc->len=10, msg_start = 5
iteration 2 - acc->len=5, msg_start = 10

iteration 2 remaining = -5 = (as this is size_t) 18446744073709551611

now the Q - how come you didn't get to this state in your tests? :)